src/ImageSharp/IO/ChunkedMemoryStream.cs

```diff
@@ -507,48 +402,82 @@ private void EnsureNotDisposed()
+    private sealed class MemoryChunkBuffer : IEnumerable<MemoryChunk>, IDisposable
+    {
+        private readonly List<MemoryChunk> memoryChunks = new();
+        private readonly MemoryAllocator allocator;
+        private readonly int allocatorCapacity;
+        private bool isDisposed;
 
-    [MethodImpl(MethodImplOptions.NoInlining)]
-    private static void ThrowInvalidSeek() => throw new ArgumentException("Invalid seek origin.");
+        public MemoryChunkBuffer(MemoryAllocator allocator)
+        {
+            this.allocatorCapacity = allocator.GetBufferCapacityInBytes();
+            this.allocator = allocator;
+        }
 
-    [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    private MemoryChunk AllocateMemoryChunk()
-    {
-        // Tweak our buffer sizes to take the minimum of the provided buffer sizes
-        // or the allocator buffer capacity which provides us with the largest
-        // available contiguous buffer size.
-        IMemoryOwner<byte> buffer = this.allocator.Allocate<byte>(Math.Min(this.allocatorCapacity, GetChunkSize(this.chunkCount++)));
+        public int ChunkCount => this.memoryChunks.Count;
+
+        public long Length { get; private set; }
 
-        return new MemoryChunk(buffer)
+        public MemoryChunk this[int index] => this.memoryChunks[index];
+
+        public void Expand()
         {
-            Next = null,
-            Length = buffer.Length()
-        };
-    }
+            IMemoryOwner<byte> buffer =
+                this.allocator.Allocate<byte>(Math.Min(this.allocatorCapacity, GetChunkSize(this.ChunkCount)));
 
-    private static void ReleaseMemoryChunks(MemoryChunk? chunk)
-    {
-        while (chunk != null)
+            MemoryChunk chunk = new(buffer)
+            {
+                Length = buffer.Length()
+            };
+
+            this.memoryChunks.Add(chunk);
+            this.Length += chunk.Length;
+        }
+
+        public void Dispose()
         {
-            chunk.Dispose();
-            chunk = chunk.Next;
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
         }
-    }
 
-    [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    private static int GetChunkSize(int i)
-    {
-        // Increment chunks sizes with moderate speed, but without using too many buffers from the same ArrayPool bucket of the default MemoryAllocator.
-        // https://github.com/SixLabors/ImageSharp/pull/2006#issuecomment-1066244720
-#pragma warning disable IDE1006 // Naming Styles
-        const int _128K = 1 << 17;
-        const int _4M = 1 << 22;
-        return i < 16 ? _128K * (1 << (int)((uint)i / 4)) : _4M;
-#pragma warning restore IDE1006 // Naming Styles
+        public IEnumerator<MemoryChunk> GetEnumerator()
+            => ((IEnumerable<MemoryChunk>)this.memoryChunks).GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator()
+            => ((IEnumerable)this.memoryChunks).GetEnumerator();
```

Comment:
I don't see any code enumerating this with `foreach`, so `IEnumerable` implementation can be deleted.